### PR TITLE
[INTERNAL] Proper release file handles when destructing File<>

### PR DESCRIPTION
**Pull request by marehr**
Attempts to fix Issue #1557

### CODE REVIEW

**Comment by marehr** [comment]
Hmm. Now some other test cases fail:
```
TEST SUITE test_index
SEQAN_ENABLE_DEBUG == 1
SEQAN_ENABLE_TESTING == 1
SEQAN_CXX_FLAGS == "SEQAN_CXX_FLAGS_NOT_SET"
SEQAN_ASYNC_IO == 1
testIndexCrossCompareDna OK
C:\nightly\co-git-develop\seqan-Nightly-64\include\seqan/file/file_mapping.h:321 FAILED!  (CloseHandle failed in unmap: "No error")
C:\nightly\co-git-develop\seqan-Nightly-64\include\seqan/file/file_mapping.h:321 FAILED!  (CloseHandle failed in unmap: "No error")
```

**Comment by h-2** [comment]
Hm, I like that is says "no error". Windows FTW!

**Comment by marehr** [comment]
So I found the error. You should always initialize your variables :disappointed:

**Comment by h-2** [comment]
I told you it might be uninitialized variables :wink:

File: include/seqan/file/file_mapping.h [hunk 1]
```diff
@@ -318,7 +318,20 @@ _unmapFile(FileMapping<TSpec> &mapping)
...
+                NULL,
+                GetLastError(),
+                MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), // Default language
+                (LPTSTR) &lpMsgBuf,
+                0,
+                NULL);
+
+            SEQAN_FAIL("CloseHandle failed in unmap: \"%s\"", lpMsgBuf /*strerror(GetLastError())*/);
+            LocalFree(lpMsgBuf);
+        }
```
Comment (h-2) [inline include/seqan/file/file_mapping.h]: Is this necessary, now that it is clear that the error came from unitialized variables?

File: include/seqan/system/file_async.h [hunk 2]
```diff
@@ -81,10 +81,16 @@ namespace seqan
...
-            handle(INVALID_HANDLE_VALUE) {}
+            handle(INVALID_HANDLE_VALUE),
+            handleAsync(INVALID_HANDLE_VALUE) {}
 
         File(void *): // to be compatible with the FILE*(NULL) constructor
-            handle(INVALID_HANDLE_VALUE) {}
+            handle(INVALID_HANDLE_VALUE),
+            handleAsync(INVALID_HANDLE_VALUE) {}
+
+        virtual ~File() {
```
Comment (h-2) [inline include/seqan/system/file_async.h]: please adhere to our coding guidelines

File: include/seqan/system/file_sync.h [hunk 3]
```diff
@@ -97,6 +97,10 @@ namespace seqan
         File(void * /*dummy*/ = NULL): // to be compatible with the FILE*(NULL) constructor
             handle(-1) {}
 
+        virtual ~File() {
```
Comment (h-2) [inline include/seqan/system/file_sync.h]: please adhere to our coding guidelines

File: include/seqan/system/file_sync.h [hunk 4]
```diff
@@ -237,7 +241,9 @@ namespace seqan
 
     ///File(int posixHandle) : handle(posixHandle) {}
 
-        virtual ~File() {}
+        virtual ~File() {
```
Comment (h-2) [inline include/seqan/system/file_sync.h]: please adhere to our coding guidelines

File: include/seqan/file/file_mapping.h (hunk 1 above)
Comment (marehr) [inline include/seqan/file/file_mapping.h]: The error message was wrong, after I changed it to this I got the correct one. It helped me to understand where the problem was.

File: include/seqan/system/file_async.h [hunk 5]
```diff
@@ -56,245 +56,277 @@
  */
 
 
-namespace seqan
-{
+namespace seqan {
```
Comment (h-2) [inline include/seqan/system/file_async.h]: I have just updated the uncrustify.cfg to to not do this.

File: include/seqan/system/file_async.h [hunk 6]
```diff
@@ -628,125 +673,134 @@ namespace seqan
...
+    typedef File<Sync<TSpec> >  Base;
 
-        typedef off_t            FilePtr;
-        typedef off_t           SizeType;   // type of file size
-        typedef size_t          SizeType_;  // type of transfer size (for read or write)
-        typedef int                Handle;
+    typedef off_t            FilePtr;
+    typedef off_t           SizeType;       // type of file size
+    typedef size_t          SizeType_;      // type of transfer size (for read or write)
+    typedef int                Handle;
```
Comment (h-2) [inline include/seqan/system/file_async.h]: please fix alignment of variables

File: include/seqan/system/file_sync.h [hunk 7]
```diff
@@ -69,399 +69,425 @@
 
 //////////////////////////////////////////////////////////////////////////////
 
-namespace seqan
-{
+namespace seqan {
```
Comment (h-2) [inline include/seqan/system/file_sync.h]: see above

File: include/seqan/system/file_sync.h [hunk 8]
```diff
@@ -69,399 +69,425 @@
...
-    public:
+public:
 
-        typedef int64_t            FilePtr;
-        typedef int64_t         SizeType;   // type of file size
-        typedef unsigned int    SizeType_;  // type of transfer size (for read or write)
-        typedef int                Handle;
+    typedef int64_t            FilePtr;
+    typedef int64_t         SizeType;       // type of file size
+    typedef unsigned int    SizeType_;      // type of transfer size (for read or write)
```
Comment (h-2) [inline include/seqan/system/file_sync.h]: see above

File: include/seqan/system/file_sync.h [hunk 9]
```diff
@@ -69,399 +69,425 @@
...
-    public:
+public:
 
-        typedef off_t            FilePtr;
-        typedef off_t           SizeType;   // type of file size
-        typedef size_t          SizeType_;  // type of transfer size (for read or write)
-        typedef int                Handle;
+    typedef off_t            FilePtr;
+    typedef off_t           SizeType;       // type of file size
+    typedef size_t          SizeType_;      // type of transfer size (for read or write)
```
Comment (h-2) [inline include/seqan/system/file_sync.h]: align variables

File: include/seqan/file/file_mapping.h (hunk 1 above)
Comment (h-2) [inline include/seqan/file/file_mapping.h]: ok

**Comment by h-2** [comment]
thanks for uncrustifying

File: include/seqan/system/file_async.h (hunk 5 above)
Comment (marehr) [inline include/seqan/system/file_async.h]: where?

File: include/seqan/system/file_async.h (hunk 6 above)
Comment (marehr) [inline include/seqan/system/file_async.h]: done

File: include/seqan/system/file_async.h (hunk 5 above)
Comment (marehr) [inline include/seqan/system/file_async.h]: done

File: include/seqan/system/file_sync.h (hunk 8 above)
Comment (marehr) [inline include/seqan/system/file_sync.h]: done

File: include/seqan/system/file_sync.h (hunk 9 above)
Comment (marehr) [inline include/seqan/system/file_sync.h]: done

File: include/seqan/system/file_sync.h (hunk 7 above)
Comment (marehr) [inline include/seqan/system/file_sync.h]: done

**Pull request merged**